src/cpp/fastrtps_deprecated/publisher/PublisherImpl.cpp

```diff
@@ -246,6 +254,99 @@ bool PublisherImpl::create_new_change_with_params(
     return false;
 }
 
+InstanceHandle_t PublisherImpl::register_instance(
+        void* instance)
+{
+    /// Preconditions
+    if (instance == nullptr)
+    {
+        logError(PUBLISHER, "Data pointer not valid");
+        return c_InstanceHandle_Unknown;
+    }
+
+    if (m_att.topic.topicKind == NO_KEY)
+    {
+        logError(PUBLISHER, "Topic is NO_KEY, operation not permitted");
+        return c_InstanceHandle_Unknown;
+    }
+
+    InstanceHandle_t instance_handle = c_InstanceHandle_Unknown;
+    bool is_key_protected = false;
+#if HAVE_SECURITY
+    is_key_protected = mp_writer->getAttributes().security_attributes().is_key_protected;
+#endif
+    mp_type->getKey(instance, &instance_handle, is_key_protected);
+
+    // Block lowlevel writer
+    auto max_blocking_time = std::chrono::steady_clock::now() +
+            std::chrono::microseconds(::TimeConv::Time_t2MicroSecondsInt64(m_att.qos.m_reliability.max_blocking_time));
+
+#if HAVE_STRICT_REALTIME
+    std::unique_lock<RecursiveTimedMutex> lock(mp_writer->getMutex(), std::defer_lock);
+    if (lock.try_lock_until(max_blocking_time))
+#else
+    std::unique_lock<RecursiveTimedMutex> lock(mp_writer->getMutex());
+#endif
+    {
+        if (m_history.register_instance(instance_handle, lock, max_blocking_time))
+        {
+            return instance_handle;
+        }
+    }
+
+    return c_InstanceHandle_Unknown;
+}
+
+bool PublisherImpl::unregister_instance(
+        void* instance,
+        const InstanceHandle_t& handle,
+        bool dispose)
+{
+    /// Preconditions
+    if (instance == nullptr)
+    {
+        logError(PUBLISHER, "Data pointer not valid");
+        return false;
+    }
+
+    if (m_att.topic.topicKind == NO_KEY)
+    {
+        logError(PUBLISHER, "Topic is NO_KEY, operation not permitted");
+        return false;
+    }
+
+    bool returned_value = false;
+    InstanceHandle_t ih = handle;
+
+#if !defined(NDEBUG)
+    if (c_InstanceHandle_Unknown == ih)
+#endif
+    {
+        bool is_key_protected = false;
+#if HAVE_SECURITY
+        is_key_protected = mp_writer->getAttributes().security_attributes().is_key_protected;
+#endif
+        mp_type->getKey(instance, &ih, is_key_protected);
+    }
+
+#if !defined(NDEBUG)
+    if (c_InstanceHandle_Unknown != handle && ih != handle)
+    {
+        logError(PUBLISHER, "handle differs from data's key.");
+        return false;
+    }
+#endif
+
+    if (m_history.key_is_registered(ih))
+    {
+        WriteParams wparams;
+        ChangeKind_t change_kind = dispose ? NOT_ALIVE_DISPOSED : NOT_ALIVE_UNREGISTERED;
+        returned_value = create_new_change_with_params(change_kind, instance, wparams, ih);
+    }
+
+    return returned_value;
```

Comment:
Same comments as on `DataWriterImpl` apply here.

Reply:
This function is common logic for `unregister_instance` and `dispose`. In case of `unregister_instance`, a NOT_ALIVE_UNREGISTERED is sent. In case of `dispose` a NOT_ALIVE_DISPOSED is sent. This old API doesn't have writer_data_lifecicle, then we cannot send NOT_ALIVE_DISPOSED_UNREGISTERED.